Introduce heartbeats (#190)
open-telemetry/opamp-spec: specification.md

OpAMP supports the following functionality:

> - Remote configuration of the Agents.

Reviewer: FYI, your editor changed `*` to `-` everywhere, which makes the diff larger than it should be.
```suggestion
* Remote configuration of the Agents.
```

Author: yeah :( i couldn't figure out how it decided what to change here... i can change it back if it really matters

Reviewer: I don't have strong opinions, but I generally try to leave the file how I found it.

The default polling interval when the Agent does not have anything to deliver is 30
seconds. This polling interval SHOULD be configurable on the Client.

> If the server sets OpAMPConnectionSettings.heartbeat_interval_seconds, the client MUST

Reviewer 2: This section applies to http transport only. I think we need to add something in the websocket section too.

    ReportsHealth                  = 0x00000800;
    // The Agent will report RemoteConfig status via AgentToServer.remote_config_status field.
    ReportsRemoteConfig            = 0x00001000;
    // The Agent can report heartbeats on a default interval of 30s.

>     // This is specified by the ServerToAgent.OpAMPConnectionSettings.heartbeat_interval_seconds field.

Reviewer 3: ```suggestion
    // The heartbeat interval is specified by the ServerToAgent.OpAMPConnectionSettings.heartbeat_interval_seconds field.
```
Maybe a little redundant in wording, but I think it may help to specify what "this" is.

    ReportsHealth                  = 0x00000800;
    // The Agent will report RemoteConfig status via AgentToServer.remote_config_status field.
    ReportsRemoteConfig            = 0x00001000;
    // The Agent can report heartbeats.
    // This is specified by the ServerToAgent.OpAMPConnectionSettings.heartbeat_interval_seconds field.
    // If this capability is true, but the Server does not set a heartbeat_interval_seconds field, the

>     // Agent should use its own configured interval, which by default will be 30s, the Server may not know this

Reviewer 3: ```suggestion
    // Agent should use its own configured interval, which by default should be 30s, the Server may not know this
```

>     // Agent should use its own configured interval, which by default should be 30s, the Server may not know this

Reviewer 2: ```suggestion
    // Agent should use its own configured interval, which by default should be 30s, the Server may not know the configured interval
```

    // know the configured interval and should not make assumptions about it.

>     // Status: [Beta]

Reviewer 2: ```suggestion
    // Status: [Development]
```

    string destination_endpoint = 1;
    Headers headers = 2;
    TLSCertificate certificate = 3;

>     optional uint64 heartbeat_interval_seconds = 4;

Reviewer 4: This is potentially an overkill/nit, but do you think we should use a [duration](https://protobuf.dev/reference/protobuf/google.protobuf/#duration) as the type?

We already have another field that is a duration but uses different units (`retry_after_nanoseconds`) which is potentially confusing. This is also in-line with [Google's AIPs](https://google.aip.dev/142).

This field is optional: if omitted the client SHOULD NOT use a client-side certificate.
This field can be used to perform a client certificate revocation/rotation.

##### OpAMPConnectionSettings.heartbeat_interval_seconds

The Client should use the offered heartbeat interval to periodically send an AgentToServer
message. At a minimum the instance_uid field MUST be set. It is recommended that the Agent
also set ComponentHealth as well. An HTTP based-client MUST use the heartbeat interval as
its polling interval.

> A heartbeat is used to keep a load balancer connection active AND inform the server that

Reviewer: Same here, does "AND" need to be in caps?
```suggestion
A heartbeat is used to keep a load balancer connection active and inform the server that
```

Author: copied 😅

the Agent is still alive and active. A server could use the heartbeat to make decisions about

> the liveness of the connected Agent.

Reviewer 5: This is a breaking change as it's written, right? Agents pre-heartbeat will ignore the heartbeat interval, the server will determine they aren't alive, then (presumably) disconnect the agent.

I think this could easily be made backwards compatible if the agent explicitly indicated that it supports this e.g. through it's capabilities.

Author: maybe not, there's guidance [here](https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#ignorable-capability-extensions). I think given we also want to use this to control the HTTP polling interval this may be a non-ignorable capability extension and would therefore require a new AgentCapability. Is there anyway around that @tigrannajaryan @andykellr ?

Reviewer 6: I think a new flag for `AgentCapabilities_ReportsHeartbeat` would be appropriate so that servers know if they should expect heartbeats.

This field is optional: if omitted the client SHOULD NOT use a client-side certificate.
This field can be used to perform a client certificate revocation/rotation.

##### OpAMPConnectionSettings.heartbeat_interval_seconds

The Client should use the offered heartbeat interval to periodically send an AgentToServer
message. At a minimum the instance_uid field MUST be set. It is recommended that the Agent
also set ComponentHealth as well. An HTTP based-client MUST use the heartbeat interval as
its polling interval.

A heartbeat is used to keep a load balancer connection active and inform the server that
the Agent is still alive and active. A server could use the heartbeat to make decisions about
the liveness of the connected Agent.

> A default of a 30s should be used if not set by the OpAMPConnectionSettings. If the

Reviewer 2: A default of 30s second polling for WebSockets seems like an overkill. What was the typical timeout you were observing?

Author: I've seen timeouts after 30s running locally and in cloud environments. Per [these AWS docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api-integration-requests.html) 30s appears to be their default.

Author: [GCP's is 5 mins ](https://cloud.google.com/run/docs/triggering/websockets#client-reconnects)

Author: [Azure recommends a 20-40 second timeout](https://learn.microsoft.com/en-us/azure/application-gateway/application-gateway-websocket#backendaddresspool-backendhttpsetting-and-routing-rule-configuration)

This field is optional: if omitted the client SHOULD NOT use a client-side certificate.
This field can be used to perform a client certificate revocation/rotation.

##### OpAMPConnectionSettings.heartbeat_interval_seconds

If the ReportsHeartbeat capability is true, the Client MUST use the offered heartbeat
interval to periodically send an AgentToServer message. At a minimum the instance_uid
field MUST be set. It is recommended that the Agent also set ComponentHealth as well.
An HTTP based-client MUST use the heartbeat interval as its polling interval.

A heartbeat is used to keep a load balancer connection active and inform the server that
the Agent is still alive and active. A server could use the heartbeat to make decisions about
the liveness of the connected Agent.

> A default of a 30s should be used if not set by the OpAMPConnectionSettings.

Reviewer 4: As brought up in #183, should we write a note about clients respecting values negotiated from the server or coming from a backpressure mechanism (eg. a Retry-After header?)

This field is optional: if omitted the client SHOULD NOT use a client-side certificate.
This field can be used to perform a client certificate revocation/rotation.

##### OpAMPConnectionSettings.heartbeat_interval_seconds

If the ReportsHeartbeat capability is true, the Client MUST use the offered heartbeat
interval to periodically send an AgentToServer message. If the capability is true
and the Server sets heartbeat_interval_seconds to 0, Agent heartbeats should be disabled.
At a minimum the instance_uid field MUST be set.
An HTTP based-client MUST use the heartbeat interval as its polling interval.

A heartbeat is used to keep a load balancer connection active and inform the server that
the Agent is still alive and active. A server could use the heartbeat to make decisions about
the liveness of the connected Agent.

> The flow for negotiating a heartbeat is described as so:

Reviewer: ```suggestion
The flow for negotiating a heartbeat is described as so:

```

This field is optional: if omitted the client SHOULD NOT use a client-side certificate.
This field can be used to perform a client certificate revocation/rotation.

##### OpAMPConnectionSettings.heartbeat_interval_seconds

If the ReportsHeartbeat capability is true, the Client MUST use the offered heartbeat
interval to periodically send an AgentToServer message. If the capability is true
and the Server sets heartbeat_interval_seconds to 0, Agent heartbeats should be disabled.

> At a minimum the instance_uid field MUST be set.

Reviewer 2: ```suggestion

Any AgentToServer message where instance_uid field is set is considered a
valid heartbeat. Note that it is not necessary to send a separate AgentToServer
message just for heartbeating purposes if another AgentToServer message
containing other data was just sent. The Agent must count heartbeating interval
from the last AgentToServer message sent.

```

> An HTTP based-client MUST use the heartbeat interval as its polling interval.

Reviewer 3: ```suggestion
An HTTP-based client MUST use the heartbeat interval as its polling interval.
```
Nit

the liveness of the connected Agent.

The flow for negotiating a heartbeat is described as so:

```
┌──────────┐                       ┌──────────┐
│          │ (1) Connect           │          │
│          ├──────────────────────►│          │
│          │                       │          │
│          │ (2) Set Heartbeat     │          │
│          │◄──────────────────────┤          │
│          │     Interval          │          │
│          │                       │          │
│  Agent   │ (3) Send Heartbeat    │  Server  │
│          ├──────────────────────►│          │
│          │                       │          │
│          │  ... heartbeat        │          │
│          │      interval         │          │
│          │                       │          │
│          │ (4) Send Heartbeat    │          │
│          ├──────────────────────►│          │
│          │                       │          │
└──────────┘                       └──────────┘
```

1. The agent connects to the server and optionally sets the ReportsHeartbeat capability. If the Agent does NOT set this capability, no heartbeats will occur.

> 2. If the Agent sets the ReportsHeartbeat capability, the server MUST respond with either a 0 indicating that heatbeats are disabled. Otherwise, the server will set a heartbeat_interval_seconds in the OpAMPConnectionSettings message. It is recommended for servers to set a 30s interval, if desired.

Reviewer 3: ```suggestion
2. If the Agent sets the ReportsHeartbeat capability, the server MUST respond by setting an interval in the heartbeat_interval_seconds field within the OpAMPConnectionSettings message. The value can either the desired interval, or `0`, indicating that the client should not send heartbeats. 30s is the recommended default interval.
```
Feel free to update to your liking, I just thought the phrasing could use some tweaking.

Author: ```suggestion
2. If the Agent sets the ReportsHeartbeat capability, the server MUST respond by setting an interval in the heartbeat_interval_seconds field within the OpAMPConnectionSettings message. The value can either be the desired interval, or `0`, indicating that the client should not send heartbeats. 30s is the recommended default interval.
```

3. If the Agent sets the ReportsHeartbeat capability AND the server has set OpAMPConnectionSettings.heartbeat_interval_seconds, the Agent MUST send a heartbeat message for the interval set by the server.
4. The Agent will continue to send heartbeats on its configured interval while alive.

> The Agent can decide not to send heartbeats by not setting the ReportsHeartbeat capability. The Server can decide to not support heartbeats by responding with an unset (or 0) for the OpAMPConnectionSettings.heartbeat_interval_seconds.

Reviewer 3: ```suggestion
The Agent can decide not to send heartbeats by not setting the ReportsHeartbeat capability. The Server can decide to not support heartbeats by responding with a value of `0` seconds in the OpAMPConnectionSettings.heartbeat_interval_seconds field.
```
`unset` just means setting `0` here, right? `uint64` isn't an optional field, will that be translated to a special `unset` value if the server does not set it? Also, we say above that the server has to respond with a value, is there a way to mark it as unset?

Author: nope, this is a relic.

This field is optional: if omitted the client SHOULD NOT use a client-side certificate.
This field can be used to perform a client certificate revocation/rotation.

##### OpAMPConnectionSettings.heartbeat_interval_seconds

> If the ReportsHeartbeat capability is true, the Client MUST use the offered heartbeat

Reviewer 2: ```suggestion

Status: [Development]

If the ReportsHeartbeat capability is true, the Client MUST use the offered heartbeat
```

interval to periodically send an AgentToServer message. If the capability is true

> and the Server sets heartbeat_interval_seconds to 0, Agent heartbeats should be disabled.

Reviewer 3: ```suggestion
and the Server sets heartbeat_interval_seconds to 0, Agent heartbeats MUST be disabled.
```
I think if we require that the agent respects the interval set by the server, it needs to respect disabling heartbeats as well.

> At a minimum the instance_uid field MUST be set.

Reviewer 2: ```suggestion
At a minimum the `AgentToServer.instance_uid` field MUST be set in the heartbeats.
```

header to indicate when SHOULD the Client attempt to reconnect. The Client SHOULD
honour the corresponding requirements of HTTP specification.

> Note: this reconnect is separate from [heartbeats](#opampconnectionsettingsheartbeat_interval_seconds). A client should still send regular heartbeat messages if it is configured to do so.

Author: no, the second sentence states that the client should continue sending heartbeats. I will clear up the wording.

> Note: this reconnect is separate from [heartbeats](#opampconnectionsettingsheartbeat_interval_seconds). A client should still send regular heartbeat messages if it is configured to do so.

Reviewer 2: Is that desirable? If the server sends 503/429 it is probably overwhelmed. Sending heartbeats can make things worse. I think we should stop sending heartbeats for the duration specified in server's response.
